File: tests/terminal/image_resume.py
#!/usr/bin/env python3
import json
import os
from pathlib import Path
import shutil
import subprocess
import sys
import tempfile
import time

from image_harness import executable, terminal_command
from image_smoke import write_source


ROOT = Path(__file__).resolve().parents[2]


def diagnostics(log, nvim_log, state):
    log.flush()
    parts = [Path(log.name).read_text(
        encoding="utf-8", errors="replace")[-8000:]]
    for path in (nvim_log, state):
        if path.is_file():
            parts.append(path.read_text(
                encoding="utf-8", errors="replace")[-8000:])
    return "\n".join(parts)


def wait_for(path, process, timeout, description, log, nvim_log, state,
             predicate=lambda value: True):
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        if process.poll() is not None:
            raise RuntimeError(
                f"terminal exited while waiting for {description}\n"
                + diagnostics(log, nvim_log, state))
        try:
            value = json.loads(path.read_text(encoding="utf-8"))
            if value.get("phase") == "error":
                raise RuntimeError(
                    f"{description} harness failed: {value.get('error')}")
            if predicate(value):
                return value
        except (FileNotFoundError, json.JSONDecodeError):
            pass
        time.sleep(0.01)
    raise RuntimeError(
        f"timed out waiting for {description}\n"
        + diagnostics(log, nvim_log, state))


def image_pixels(image_tool, screenshot):
    pixels = subprocess.check_output([
        image_tool, str(screenshot), "-crop", "1024x768+0+0",
        "-alpha", "off", "-depth", "8", "rgb:-",
    ], cwd=ROOT, timeout=15)
    count = 0
    for offset in range(0, len(pixels) - 2, 3):
        red, green, blue = pixels[offset:offset + 3]
        if red >= 180 and green <= 100 and blue >= 180:
            count += 1
    return count


def settled_image_pixels(importer, image_tool, screenshot, timeout=3):
    deadline = time.monotonic() + timeout
    previous = None
    stable_since = None
    while time.monotonic() < deadline:
        subprocess.run([
            importer, "-window", "root", str(screenshot),
        ], cwd=ROOT, check=True, timeout=10)
        count = image_pixels(image_tool, screenshot)
        now = time.monotonic()
        if count == previous:
            stable_since = stable_since or now
            if now - stable_since >= 0.25:
                return count
        else:
            previous = count
            stable_since = now
        time.sleep(0.01)
    raise RuntimeError(
        f"terminal image pixels did not settle: last count={previous}")


def run_case(kind, terminal, nvim, importer, image_tool):
    name = f"resume-{kind}"
    artifact = ROOT / ".test-data" / f"terminal-{name}-failure.png"
    artifact.unlink(missing_ok=True)
    with tempfile.TemporaryDirectory(
            prefix=f"neoagent-image-{name}-") as temporary:
        directory = Path(temporary)
        workspace = directory / "workspace"
        persistence = directory / "state"
        workspace.mkdir()
        persistence.mkdir()
        source = directory / "source.png"
        ready = directory / "ready.json"
        state = directory / "state.json"
        done = directory / "done.json"
        stop = directory / "stop"
        socket = directory / "nvim.sock"
        screenshot = directory / "screen.png"
        terminal_log = directory / "terminal.log"
        nvim_log = directory / "nvim.log"
        write_source(source, True)
        environment = os.environ.copy()
        environment.update({
            "NEOAGENT_IMAGE_RESUME_ROOT": str(ROOT),
            "NEOAGENT_IMAGE_RESUME_READY": str(ready),
            "NEOAGENT_IMAGE_RESUME_STATE": str(state),
            "NEOAGENT_IMAGE_RESUME_DONE": str(done),
            "NEOAGENT_IMAGE_RESUME_STOP": str(stop),
            "NEOAGENT_IMAGE_RESUME_SOURCE": str(source),
            "NEOAGENT_IMAGE_RESUME_PERSISTENCE": str(persistence),
            "NEOAGENT_IMAGE_RESUME_WORKSPACE": str(workspace),
        })
        child = [
            nvim, "-u", "NONE", "-i", "NONE", "-n",
            "--listen", str(socket),
            "-V1" + str(nvim_log),
            "-c", "luafile " + str(ROOT / "tests/terminal/image_resume.lua"),
        ]
        command = terminal_command(kind, terminal, child, directory)
        with terminal_log.open("wb") as log:
            process = subprocess.Popen(
                command, cwd=workspace, env=environment,
                stdout=log, stderr=log)
            try:
                def send(keys):
                    subprocess.run([
                        nvim, "--server", str(socket),
                        "--remote-send", keys,
                    ], cwd=ROOT, check=True, timeout=5)

                draft = wait_for(
                    state, process, 15, f"{name} open draft",
                    log, nvim_log, state,
                    lambda value: value.get("phase") == "draft"
                    and value.get("current_window")
                    == value.get("input_window"))
                if draft.get("image") is not None:
                    raise RuntimeError(
                        f"{name} draft unexpectedly contained an image: "
                        f"{draft}")
                send("<M-r>")
                wait_for(
                    state, process, 15, f"{name} resume picker",
                    log, nvim_log, state,
                    lambda value: value.get("presentation_window") is not None
                    and value.get("current_window")
                    == value.get("presentation_window"))
                send("<CR>")
                resumed = wait_for(
                    ready, process, 15, f"{name} resumed transcript",
                    log, nvim_log, state)
                if resumed.get("current_window") != resumed.get("input_window"):
                    raise RuntimeError(
                        f"{name} did not resume with composer focus: {resumed}")
                send("<M-k>")
                navigated = wait_for(
                    done, process, 15, f"{name} Alt+K navigation",
                    log, nvim_log, state)
                if navigated.get("current_window") \
                        != navigated.get("transcript_window"):
                    raise RuntimeError(
                        f"{name} Alt+K did not focus transcript: {navigated}")
                position = navigated.get("image_position") or {}
                if position.get("row") != 0:
                    raise RuntimeError(
                        f"{name} fixture image is semantically visible: "
                        f"{navigated}")
                count = settled_image_pixels(
                    importer, image_tool, screenshot)
                placements = navigated.get("placements") or []
                if placements or count >= 1000:
                    artifact.parent.mkdir(parents=True, exist_ok=True)
                    shutil.copyfile(screenshot, artifact)
                    raise RuntimeError(
                        f"{name} placed an off-screen transcript image over "
                        f"the selected card: pixels={count}, "
                        f"placements={placements}; capture: {artifact}")
                stop.write_text("stop\n", encoding="utf-8")
                process.wait(timeout=5)
                if process.returncode != 0:
                    raise RuntimeError(
                        f"{name} Neovim failed\n"
                        + diagnostics(log, nvim_log, state))
                print(f"PASS {name}: Alt+K retained no off-screen image pixels")
            finally:
                if process.poll() is None:
                    process.terminate()
                    try:
                        process.wait(timeout=2)
                    except subprocess.TimeoutExpired:
                        process.kill()
                        process.wait(timeout=2)


def main():
    if os.environ.get("NEOAGENT_IMAGE_RESUME_XVFB") != "1":
        xvfb = executable(os.environ.get("XVFB_RUN", "xvfb-run"))
        if not xvfb:
            print("SKIP resumed terminal images: xvfb-run is unavailable")
            return 0
        environment = os.environ.copy()
        environment["NEOAGENT_IMAGE_RESUME_XVFB"] = "1"
        return subprocess.run([
            xvfb, "-a", "--server-args=-screen 0 1024x768x24",
            sys.executable, str(Path(__file__).resolve()),
        ], cwd=ROOT, env=environment).returncode

    nvim = executable(os.environ.get("NEOAGENT_NVIM", "nvim"))
    importer = executable(os.environ.get("IMPORT", "import"))
    image_tool = executable(os.environ.get("MAGICK", "convert"))
    missing = [name for name, value in (
        ("nvim", nvim), ("import", importer),
        ("ImageMagick", image_tool),
    ) if not value]
    if missing:
        print("SKIP resumed terminal images: unavailable "
              + ", ".join(missing))
        return 0

    ran = 0
    kitty = executable(os.environ.get("KITTY", "kitty"))
    if kitty:
        run_case("kitty", kitty, nvim, importer, image_tool)
        ran += 1
    else:
        print("SKIP resumed Kitty image: Kitty is unavailable")
    konsole = executable(os.environ.get("KONSOLE", "konsole"))
    if konsole:
        run_case("konsole", konsole, nvim, importer, image_tool)
        ran += 1
    else:
        print("SKIP resumed Konsole image: Konsole is unavailable")
    if not ran:
        print("SKIP resumed terminal images: no supported terminal is installed")
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except (OSError, RuntimeError, subprocess.SubprocessError) as error:
        print(f"FAIL resumed terminal images: {error}", file=sys.stderr)
        raise SystemExit(1)

File: tests/terminal/image_resume.lua
local root = assert(vim.env.NEOAGENT_IMAGE_RESUME_ROOT)
vim.opt.runtimepath:prepend(root)
vim.opt.termguicolors = true

local ready_path = assert(vim.env.NEOAGENT_IMAGE_RESUME_READY)
local state_path = assert(vim.env.NEOAGENT_IMAGE_RESUME_STATE)
local done_path = assert(vim.env.NEOAGENT_IMAGE_RESUME_DONE)
local stop_path = assert(vim.env.NEOAGENT_IMAGE_RESUME_STOP)
local source_path = assert(vim.env.NEOAGENT_IMAGE_RESUME_SOURCE)
local persistence = assert(vim.env.NEOAGENT_IMAGE_RESUME_PERSISTENCE)
local workspace = assert(vim.env.NEOAGENT_IMAGE_RESUME_WORKSPACE)

local function publish(path, value)
  local temporary = path .. ".tmp"
  assert(vim.fn.writefile({ vim.json.encode(value) }, temporary) == 0)
  assert(vim.uv.fs_rename(temporary, path))
end

local function read_file(path)
  local descriptor = assert(vim.uv.fs_open(path, "r", 0))
  local stat = assert(vim.uv.fs_fstat(descriptor))
  local data = assert(vim.uv.fs_read(descriptor, stat.size, 0))
  assert(vim.uv.fs_close(descriptor))
  return data
end

local function assistant(content, stop_reason)
  return {
    role = "assistant",
    content = content,
    api = "fake-api",
    provider = "fake",
    model = "test",
    stopReason = stop_reason or "stop",
    timestamp = 2,
    usage = {
      input = 0,
      output = 0,
      cacheRead = 0,
      cacheWrite = 0,
      totalTokens = 0,
      cost = {
        input = 0,
        output = 0,
        cacheRead = 0,
        cacheWrite = 0,
        total = 0,
      },
    },
  }
end

local session = assert(require("neoagent.profile_sessions").new({
  profile_id = "neo",
  workspace = workspace,
  persistence = { enabled = true, directory = persistence },
}))
assert(session:append({
  role = "user",
  content = "Inspect the persisted image fixture.",
  timestamp = 1,
}, {
  model = { provider = "fake", model = "test" },
  thinking_level = "high",
}))
assert(session:append(assistant({
  { type = "thinking", thinking = "I will inspect the image." },
  {
    type = "toolCall",
    id = "read-image",
    name = "read_file",
    arguments = { path = source_path },
  },
}, "toolUse")))
assert(session:append({
  role = "toolResult",
  toolCallId = "read-image",
  toolName = "read_file",
  isError = false,
  timestamp = 3,
  content = {
    { type = "text", text = "Read image file [image/png]" },
    {
      type = "image",
      mimeType = "image/png",
      data = vim.base64.encode(read_file(source_path)),
    },
  },
}))
assert(session:append(assistant({
  {
    type = "thinking",
    thinking = table.concat({
      "The image is followed by enough persisted transcript content",
      "to place it outside the viewport when the last card is selected.",
      "This mirrors a resumed tool-result image followed by analysis.",
      "The terminal placement must follow the final transcript viewport.",
    }, "\n\n"),
  },
  {
    type = "toolCall",
    id = "inspect-image",
    name = "shell",
    arguments = { command = "identify fixture.png" },
  },
}, "toolUse")))
assert(session:append({
  role = "toolResult",
  toolCallId = "inspect-image",
  toolName = "shell",
  isError = false,
  timestamp = 4,
  content = { {
    type = "text",
    text = table.concat({
      "fixture.png PNG 640x384",
      "The diagnostic completed successfully.",
      "The persisted result has no additional image.",
      "Rendering can continue with the final response.",
    }, "\n"),
  } },
}))
assert(session:append(assistant({
  {
    type = "thinking",
    thinking = table.concat({
      "I can now summarize the fixture.",
      "The earlier image belongs above this final card.",
      "Its terminal pixels must not cover this text after navigation.",
    }, "\n\n"),
  },
  {
    type = "text",
    text = table.concat({
      "The fixture uses a deliberately distinctive palette.",
      "It has a wide landscape aspect ratio.",
      "The bright pixels make stale terminal placement measurable.",
      "This final response is the card selected from the composer.",
      "The original image is semantically higher in the transcript.",
    }, "\n\n"),
  },
})))

local model = {
  api = "fake-api",
  provider = "fake",
  id = "test",
  input = { "text", "image" },
}
function model:stream()
  error("the image resume harness must not make model requests", 0)
end

local errors = {}
local neoagent = require("neoagent")
local applet = neoagent.setup({
  workspace_trust = false,
  default_registry = false,
  default_model = { provider = "fake", model = "test" },
  providers = {
    fake = { api = "fake-api", models = { test = {
      input = { "text", "image" },
    } } },
  },
  _apis = { ["fake-api"] = function() return model end },
  persistence = {
    enabled = true,
    directory = persistence,
    workspace_settings = false,
  },
  tools = {},
  agent_instructions = false,
  skills = false,
  ui = {
    position = "center",
    style = "codex",
    show_thinking = true,
    images = { display = "always" },
  },
})
assert(applet:open())
assert(applet:is_open())

local view = assert(applet:view())
local pane = assert(view.transcript.pane)
local images = assert(view.image_system)
local backend = assert(images.backend)
local navigated = false
local action_started
local stable_since
local finishing = false
local last_state = 0

local function current_placements()
  local owner = backend.owners and backend.owners[pane]
  local result = {}
  for _, placement in ipairs(owner and owner.placements or {}) do
    result[#result + 1] = {
      key = placement.key,
      geometry = vim.deepcopy(placement.geometry),
    }
  end
  return result
end

local function snapshot(phase)
  local native = pane:native()
  local window = native and native.window
  local input = view:pane("input")
  local input_native = input and input:native()
  local presentation = view:pane("presentation-filter")
    or view:pane("presentation")
  local presentation_native = presentation and presentation:native()
  local image_key, image = next(pane.layout and pane.layout.images or {})
  local position
  if window and image then
    local line = vim.api.nvim_buf_get_lines(
      native.buffer, image.row, image.row + 1, false)[1] or ""
    position = vim.fn.screenpos(window, image.row + 1,
      require("applet.util").byte_col(line, image.col) + 1)
  end
  return {
    phase = phase,
    errors = vim.deepcopy(errors),
    mode = vim.api.nvim_get_mode().mode,
    current_window = vim.api.nvim_get_current_win(),
    transcript_window = window,
    input_window = input_native and input_native.window,
    presentation_window = presentation_native and presentation_native.window,
    image_key = image_key,
    image = image and {
      row = image.row,
      col = image.col,
      width = image.width,
      height = image.height,
    } or nil,
    image_position = position,
    placements = current_placements(),
    view = window and vim.api.nvim_win_call(window, vim.fn.winsaveview) or nil,
    cursor = window and vim.api.nvim_win_get_cursor(window) or nil,
    image_stats = images:_stats(),
    output_pending = backend.output_operation ~= nil,
    pending_presentations = backend.pending and vim.tbl_count(backend.pending) or 0,
  }
end

local function current_phase()
  if navigated then return "navigated" end
  if next(pane.layout and pane.layout.images or {}) then return "resumed" end
  if view:pane("presentation-filter") or view:pane("presentation") then
    return "selection"
  end
  return "draft"
end

local timer = vim.uv.new_timer()
local started = vim.uv.hrtime()
timer:start(5, 5, vim.schedule_wrap(function()
  if finishing then return end
  local ok, err = xpcall(function()
    if #errors > 0 then error(table.concat(errors, "\n"), 0) end
    local current_view = applet:view()
    if current_view and current_view ~= view then
      view = current_view
      pane = assert(view.transcript.pane)
      images = assert(view.image_system)
      backend = assert(images.backend)
    end
    local state = snapshot(current_phase())
    local now = vim.uv.hrtime()
    if now - last_state >= 20 * 1e6 then
      last_state = now
      publish(state_path, state)
    end
    if state.image and state.current_window == state.input_window
        and vim.fn.filereadable(ready_path) ~= 1 then
      publish(ready_path, state)
    end
    if not navigated and state.image
        and state.current_window == state.transcript_window then
      navigated = true
      action_started = now
      state.phase = "navigated"
      stable_since = nil
    end
    if navigated then
      local settled = state.current_window == state.transcript_window
        and state.image_stats.pending_preparations == 0
        and not state.output_pending
        and state.pending_presentations == 0
      if settled then
        stable_since = stable_since or now
      else
        stable_since = nil
      end
      if stable_since and now - stable_since >= 250 * 1e6
          and vim.fn.filereadable(done_path) ~= 1 then
        state.phase = "navigated"
        publish(done_path, state)
      end
    end
    if vim.fn.filereadable(stop_path) == 1 then
      finishing = true
      timer:stop()
      timer:close()
      vim.schedule(function() vim.cmd("qa!") end)
      return
    end
    if action_started and now - action_started > 10 * 1e9 then
      error("resume navigation did not settle: "
        .. vim.inspect(snapshot("timeout")), 0)
    end
    if now - started > 30 * 1e9 then
      error("image resume harness timed out: "
        .. vim.inspect(snapshot("timeout")), 0)
    end
  end, debug.traceback)
  if not ok then
    finishing = true
    pcall(publish, done_path, { phase = "error", error = err })
    timer:stop()
    timer:close()
    vim.api.nvim_err_writeln(err)
    vim.schedule(function() vim.cmd("cquit 1") end)
  end
end))
